include/platform/messagepool.h

```diff
@@ -0,0 +1,100 @@
+/*
+ *  Copyright (c) 2016, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *   This file includes the platform abstraction for the message pool.
+ */
+
+#ifndef MESSAGEPOOL_H_
+#define MESSAGEPOOL_H_
+
+#include <stdint.h>
+#include <openthread-types.h>
+
+/**
+ * @defgroup messagepool MessagePool
+ * @ingroup platform
+ *
+ * @brief
+ *   This module includes the platform abstraction for the message pool.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This structure contains a pointer to the next Message buffer.
+ *
+ */
+struct BufferHeader
+{
+    struct BufferHeader *mNext;  ///< A pointer to the next Message buffer.
+};
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Initialize the platform implemented message pool with the provided buffer list.
+ *
+ * @param[in] aFreeBufferList   A linked list of free buffers.
+ * @param[in] aNumFreeBuffers   A pointer to an int containing the number of free buffers in aFreeBufferList
+ *
+ */
+void otPlatMessagePoolInit(struct BufferHeader *aFreeBufferList, int *aNumFreeBuffers);
```

Comment:
I am indeed trying to allow for an alternate implementation of MessagePool::NewBuffer which is single threaded by design.  The key to making it multithreaded is to grab some critical section protection object, like a mutex, before modifying the free list and the count of free buffers.  So on the RTOS platform that I'm using, I can craft a function that will employ this little extra protection.  The condition where I need this, is when the upper layer TCP stack wants to send a packet to openthread for transmission.  the TCP stack executes on a separate task and otSendIP6Datagram() expects an otMessage object. So that task needs to allocate an otMessage object and then copy the packet data from the TCP object to the otMessage object and queue it to the OpenThread Task for transmission via otSendIP6Datagram.  You might wonder why I can't simply queue the TCP stack's packet buffer and the answer is that the TCP stack assumes that the function call to send is synchronous. When that call returns it keeps ownership of that packet, meaning I can't queue it to another task (the OT task).  I have to instead make a copy and queue another object.  Since I have to copy it anyway to an otMessage object, I kill two birds by having an allocate function that I can call from another task.  I can totally understand why this might be hard to grasp. I'm using a 1000 words for what could probably be explained by one good picture.